core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -78,6 +78,12 @@ private[kafka] object LogValidator extends Logging {
       if (batch.isControlBatch)
         throw new InvalidRecordException("Clients are not allowed to write control records")
     }
+
+    if (batch.isTransactional && toMagic < RecordBatch.MAGIC_VALUE_V2)
+      throw new UnsupportedForMessageFormatException(s"Transactional records cannot be used with magic version $toMagic")
```

Comment:
What I'm wondering if users made a mistake in operations while setting `messageFormatVersion`, in that case should we throw an exception to stop the broker upon receiving a first transactional message?